enterprise/reporting/reporting.go

```diff
@@ -361,126 +377,116 @@ func (r *DefaultReporter) mainLoop(ctx context.Context, c types.SyncerConfig) {
+			loopStart := time.Now()
+			currentMin := time.Now().UTC().Unix() / 60
+
+			getReportsStart := time.Now()
+			reports, reportedAt, err := r.getReports(currentMin, c.ConnInfo)
+			getReportsTimer.Since(getReportsStart)
+			getReportsCount.Observe(float64(len(reports)))
+			if len(reports) == 0 {
+				if err == nil {
+					lastReportedAtTime.Store(loopStart)
+				} else {
+					r.log.Errorw("getting reports", "error", err)
+				}
+				select {
+				case <-ctx.Done():
+					r.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
+					return nil
+				case <-time.After(r.sleepInterval.Load()):
+				}
 				continue
 			}
-			metricToSend := metric
-			requestChan <- struct{}{}
-			if errCtx.Err() != nil {
-				// if any of errGroup's goroutines fail - don't send anymore requests for this batch
-				break
-			}
-			errGroup.Go(func() error {
-				err := r.sendMetric(errCtx, netClient, c.Label, metricToSend)
-				<-requestChan
-				return err
-			})
-		}
 
-		err = errGroup.Wait()
-		if err == nil {
-			dbHandle, err := r.getDBHandle(c.ConnInfo)
-			if err != nil {
-				panic(err)
-			}
-			_, err = dbHandle.Exec(`DELETE FROM `+ReportsTable+` WHERE reported_at = $1`, reportedAt)
-			if err != nil {
-				r.log.Errorf(`[ Reporting ]: Error deleting local reports from %s: %v`, ReportsTable, err)
+			lastReportedAtTime.Store(time.Unix(reportedAt*60, 0))
+			getAggregatedReportsStart := time.Now()
+			metrics := r.getAggregatedReports(reports)
+			getAggregatedReportsTimer.Since(getAggregatedReportsStart)
+			getAggregatedReportsCount.Observe(float64(len(metrics)))
+
+			errGroup, errCtx := errgroup.WithContext(ctx)
+			// default to -1 to allow unlimited concurrency
+			errGroup.SetLimit(config.GetInt("Reporting.maxConcurrentRequests", -1))
+			for _, metric := range metrics {
+				if r.whActionsOnly && metric.SourceCategory != "warehouse" {
+					// if whActionsOnly is true, we only send reports for wh actions sources
+					// we silently drop all other reports
+					continue
+				}
+				metricToSend := metric
+				requestChan <- struct{}{}
+				if errCtx.Err() != nil {
+					// if any of errGroup's goroutines fail - don't send anymore requests for this batch
+					break
+				}
+				errGroup.Go(func() error {
+					err := r.sendMetric(errCtx, netClient, c.Label, metricToSend)
+					<-requestChan
+					return err
+				})
 			}
 
-			vacuumStart := time.Now()
-			var sizeEstimate int64
-			if err := dbHandle.QueryRowContext(
-				ctx,
-				fmt.Sprintf(`SELECT pg_table_size(oid) from pg_class where relname='%s';`, ReportsTable),
-			).Scan(&sizeEstimate); err != nil {
-				r.log.Errorn(
-					`[ Reporting ]: Error getting table size estimate`,
-					logger.NewErrorField(err),
-				)
-			}
-			if sizeEstimate > config.GetInt64("Reporting.vacuumThresholdBytes", 5*bytesize.GB) {
-				if _, err := dbHandle.ExecContext(ctx, `vacuum full analyze reports;`); err != nil {
+			err = errGroup.Wait()
+			if err == nil {
```

Comment:
wouldn't be more readable if you handled the err case first?

```suggestion
			err = errGroup.Wait()
			if err != nil {
```
hard to read the code in github diff, but I understand we do nothing if there is an error. Is this correct? Maybe we should log it ?

Reply:
Yes, we should log it. 

By the way, you can change the GitHub diff settings to `hide whitespace`, changes are much more clear that way